components/console/esp_linenoise.h

```diff
@@ -0,0 +1,237 @@
+ * 
+ * @returns - ESP_OK on success
+ *          - ESP_INVALID_ARG if the handle is invalid
+ */
+esp_err_t  esp_linenoise_set_free_hints_callback(esp_linenoise_handle_t handle, esp_linenoise_free_hints_callback_t *callback);
+
+/**
+ * @brief This function is used by the callback function registered by the user
+ * in order to add completion options given the input string when the
+ * user typed <tab>.
+ * 
+ * @param lc the linenoise completion object that we are augmenting
+ * @param str the tab completion to add
+ * 
+ * @returns - ESP_OK on success
+ *          - ESP_NO_MEM if we ran out of memory
+ */
+esp_err_t  esp_linenoise_add_completion(esp_linenoise_completions_t *lc, const char *str);
+
+/**
+ * @brief This is the API call to add a new entry in the linenoise history.
+ * It uses a fixed array of char pointers that are shifted (memmoved)
+ * when the history max length is reached in order to remove the older
+ * entry and make room for the new one, so it is not exactly suitable for huge
+ * histories, but will work well for a few hundred of entries.
+ * 
+ * @returns - ESP_OK on success, or the line is already at the front of the history
+ *          - ESP_ERR_NOT_SUPPORTED if the history max length is zero
+ *          - ESP_ERR_NO_MEM if we ran out of memory
+ *          - ESP_INVALID_ARG if the handle is invalid, or line is NULL
+ */
+esp_err_t  esp_linenoise_history_add(esp_linenoise_handle_t handle, const char *line);
+
+/**
+ * @brief Set the maximum length for the history. This function can be called even
+ * if there is already some history, the function will make sure to retain
+ * just the latest 'len' elements if the new history length value is smaller
+ * than the amount of items already inside the history.
+ * 
+ * @returns - ESP_OK on success
+ *          - ESP_ERR_NO_MEM if we ran out of memory
+ *          - ESP_INVALID_ARG if the handle is invalid, or length < 1
+ */
+esp_err_t  esp_linenoise_history_set_max_length(esp_linenoise_handle_t handle, int length);
+
+/**
+ * @brief Save the history in the specified file.
+ * 
+ * @returns - ESP_OK on success
+ *          - ESP_INVALID_ARG if the handle is invalid, or filename is NULL
+ */
+esp_err_t  esp_linenoise_history_save(esp_linenoise_handle_t handle, const char *filename);
+
+/**
+ * @brief Load the history from the specified file.
+ * 
+ * @returns - ESP_OK on success
+ *          - ESP_FAIL if the file does not exist
+ *          - ESP_ERR_NO_MEM if we ran out of memory
+ *          - ESP_INVALID_ARG if the handle is invalid, or filename is NULL
+ */
+esp_err_t  esp_linenoise_history_load(esp_linenoise_handle_t handle, const char *filename);
+
+/**
+ * @brief Free all history.
+ * 
+ * @returns - ESP_OK on success
+ *          - ESP_INVALID_ARG if the handle is invalid
+ */
+esp_err_t  esp_linenoise_history_free(esp_linenoise_handle_t handle);
+
+
+/**
+ * @brief Linenoise does not necessarily allocate with malloc. Any memory returned by
```

Comment:
This maybe was true for the original library, is this still necessary?

Reply:
I had the same thought. Ideally we should eliminate returning memory that we allocate ourself, IMO.

`esp_err_t esp_linenoise_get_line(esp_linenoise_handle_t handle, const char *prompt, char* out_line, size_t len);`